src/engine/Operation.cpp

```diff
@@ -69,6 +69,105 @@ void Operation::recursivelySetTimeConstraint(
+  }
+  // Make sure that the results that are written to the cache have the
+  // correct runtimeInfo. The children of the runtime info are already set
+  // correctly because the result was computed, so we can pass `nullopt` as
+  // the last argument.
+  if (result.isDataEvaluated()) {
+    updateRuntimeInformationOnSuccess(result.idTable().size(),
+                                      ad_utility::CacheStatus::computed,
+                                      timer.msecs(), std::nullopt);
+  } else {
+    runtimeInfo().status_ = RuntimeInformation::lazilyMaterialized;
+    result.runOnNewChunkComputed([this, isRoot](
+                                     const IdTable& idTable,
+                                     std::chrono::milliseconds duration) {
+      runtimeInfo().totalTime_ += duration;
+      runtimeInfo().originalOperationTime_ = runtimeInfo().getOperationTime();
+      runtimeInfo().numRows_ += idTable.numRows();
+      runtimeInfo().numCols_ = idTable.numColumns();
+      LOG(DEBUG) << "Computed partial chunk of size " << idTable.numRows()
+                 << " x " << idTable.numColumns() << std::endl;
+      if (isRoot) {
+        signalQueryUpdate();
+      }
+    });
+  }
+  // Apply LIMIT and OFFSET, but only if the call to `computeResult` did not
+  // already perform it. An example for an operation that directly computes
+  // the Limit is a full index scan with three variables. Note that the
+  // `QueryPlanner` does currently only set the limit for operations that
+  // support it natively, except for operations in subqueries. This means
+  // that a lot of the time the limit is only artificially applied during
+  // export, allowing the cache to reuse the same operation for different
+  // limits and offsets.
+  if (!supportsLimit()) {
+    runtimeInfo().addLimitOffsetRow(_limit, std::chrono::milliseconds{0}, true);
+    result.applyLimitOffset(_limit, [runtimeInfo = getRuntimeInfoPointer()](
+                                        std::chrono::milliseconds limitTime) {
+      runtimeInfo->totalTime_ += limitTime;
+    });
+  } else {
+    result.enforceLimitOffset(_limit);
+  }
+  return result;
+}
+
+// _____________________________________________________________________________
+CacheValue Operation::runComputationAndPrepareForCache(
+    const ad_utility::Timer& timer, ComputationMode computationMode,
+    const std::string& cacheKey, bool pinned, bool isRoot) {
+  auto& cache = _executionContext->getQueryTreeCache();
+  auto result = runComputation(timer, computationMode, isRoot);
+  result.cacheDuringConsumption(
+      [&cache](const std::optional<IdTable>& currentIdTable,
+               const IdTable& newIdTable) {
+        auto currentSize = currentIdTable.has_value()
+                               ? CacheValue::getSize(currentIdTable.value())
+                               : 0_B;
+        return cache.getMaxSizeSingleEntry() >=
+               currentSize + CacheValue::getSize(newIdTable);
```

Comment:
the call to `getMaxSizeSingle...` is expensive (global lock). 
Cache it once and reuse for every chunk.